chainstate/types/src/block_status.rs

```diff
@@ -13,48 +13,185 @@
     }
 
     /// Advance the last successful validation stage to the specified value.
     /// Note that the stage can only be advanced one step at a time.
     pub fn advance_validation_stage_to(&mut self, new_stage: BlockValidationStage) {
-        assert!(self.last_successful_validation_stage.next() == Some(new_stage));
-        self.last_successful_validation_stage = new_stage;
+        assert!(self.last_valid_stage().next() == Some(new_stage));
+        self.set_last_valid_stage(new_stage);
     }
 
-    pub fn is_valid(&self) -> bool {
-        self.last_successful_validation_stage == BlockValidationStage::FullyChecked
+    pub fn last_valid_stage(&self) -> BlockValidationStage {
+        let val = self.0.get_field(BlockStatusField::ValidationStage);
+        BlockValidationStage::from_u64(val).expect("Corrupted BlockValidationStage")
     }
 
-    pub fn last_valid_stage(&self) -> BlockValidationStage {
-        self.last_successful_validation_stage
+    // Note: it's better to keep this function private if possible.
+    fn set_last_valid_stage(&mut self, stage: BlockValidationStage) {
+        self.0.set_field(
+            BlockStatusField::ValidationStage,
+            stage as BlockStatusEffectiveType,
+        );
+    }
+
+    pub fn is_fully_valid(&self) -> bool {
+        self.last_valid_stage() == BlockValidationStage::FullyChecked && self.is_ok()
+    }
+
+    pub fn is_ok(&self) -> bool {
+        !(self.validation_failed() || self.has_invalid_parent())
+    }
+
+    pub fn set_validation_failed(&mut self) {
+        self.0.set_flag(BlockStatusField::ValidationFailedBit, true);
+    }
+
+    pub fn validation_failed(&self) -> bool {
+        self.0.get_flag(BlockStatusField::ValidationFailedBit)
+    }
+
+    pub fn set_has_invalid_parent(&mut self) {
+        self.0.set_flag(BlockStatusField::InvalidParentBit, true);
+    }
+
+    pub fn has_invalid_parent(&self) -> bool {
+        self.0.get_flag(BlockStatusField::InvalidParentBit)
+    }
+
+    // Note: this is needed for testing only.
+    pub fn reserved_bits(&self) -> BlockStatusEffectiveType {
+        self.0.get_field(BlockStatusField::ReservedArea1)
+            | self.0.get_field(BlockStatusField::ReservedArea2)
+    }
+}
+
+impl std::fmt::Display for BlockStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "BlockStatus(last_valid_stage: {}, validation_failed: {}, has_invalid_parent: {})",
+            self.last_valid_stage(),
+            self.validation_failed(),
+            self.has_invalid_parent()
+        )
+    }
+}
+
+// Each value here represents the bit number where the corresponding field starts.
+// I.e. the validation stage is in the lowest byte, flags are in the highest byte and the rest
+// is reserved for future use.
+#[derive(Sequence, Clone, Copy)]
+enum BlockStatusField {
+    ValidationStage = 0,
+    ReservedArea1 = 8,
+    ValidationFailedBit = 56,
```

Comment:
These bits seem to be arbitrary... can you decompose them to | ops and add documentation for the logic you used to build them?

Reply:
Well, there is a comment above that says the the flags go to the highest byte, so 56 is 7*8 - the starting bit of the 7th byte. But I agree that an additional comment near ValidationFailedBit itself would be nice to have.

But I'm also thinking about moving the flags to the very end of the u64, i.e. ValidationFailedBit = 62, InvalidParentBit = 63. I mean, the idea was to have a large reserved area of bits in the middle (just in case); moving them to the end will make it even larger, so it kind of makes sense.